UpdateCgnOrchestrator/index.ts

```diff
@@ -0,0 +1,158 @@
+import { ActivityInput as SendMessageActivityInput } from "../SendMessageActivity/handler";
+import {
+  ActivityInput,
+  ActivityResult
+} from "../UpdateCgnStatusActivity/handler";
+import { trackEvent, trackException } from "../utils/appinsights";
+import { getMessage } from "../utils/messages";
+import { internalRetryOptions } from "../utils/retry_policies";
+
+export const OrchestratorInput = t.interface({
+  fiscalCode: FiscalCode,
+  newStatus: CgnStatus
+});
+export type OrchestratorInput = t.TypeOf<typeof OrchestratorInput>;
+
+const NOTIFICATION_DELAY_SECONDS = 10;
+
+const trackExceptionAndThrow = (
+  context: IOrchestrationFunctionContext,
+  logPrefix: string
+) => (err: Error | t.Errors, name: string) => {
+  context.log.verbose(
+    err instanceof Error
+      ? `${logPrefix}|ERROR=${err.message}`
+      : `${logPrefix}|ERROR=${readableReport(err)}`
+  );
+  trackException({
+    exception: new Error(`${logPrefix}|ERROR=${String(err)}`),
+    properties: {
+      name
+    }
+  });
+  throw new Error(String(err));
+};
+
+const getMessageType = (cgnStatus: CgnStatus) => {
+  if (CgnRevokedStatus.is(cgnStatus)) {
+    return "CgnRevokedStatus";
+  }
+  if (CgnCanceledStatus.is(cgnStatus)) {
+    return "CgnCanceledStatus";
+  } else {
+    return "CgnActivatedStatus";
+  }
+};
+
+export const handler = function*(
+  context: IOrchestrationFunctionContext,
+  logPrefix: string = "UpdateCgnOrchestrator"
+): Generator {
+  const trackExAndThrow = trackExceptionAndThrow(context, logPrefix);
+  context.df.setCustomStatus("RUNNING");
+  const trackEventIfNotReplaying = (evt: EventTelemetry) =>
+    context.df.isReplaying ? constVoid : trackEvent(evt);
+
+  const trackExceptionIfNotReplaying = (evt: ExceptionTelemetry) =>
+    context.df.isReplaying ? constVoid : trackException(evt);
+
+  const input = context.df.getInput();
+  const decodedInput = OrchestratorInput.decode(input).getOrElseL(e =>
+    trackExAndThrow(e, "cgn.update.exception.decode.input")
+  );
+
+  const { fiscalCode, newStatus } = decodedInput;
+  const tagOverrides = {
+    "ai.operation.id": fiscalCode,
+    "ai.operation.parentId": fiscalCode
+  };
+
+  try {
+    const updateCgnStatusActivityInput = ActivityInput.encode({
+      cgnStatus: newStatus,
+      fiscalCode
+    });
+    const updateStatusResult = yield context.df.callActivityWithRetry(
+      "UpdateCgnStatusActivity",
+      internalRetryOptions,
+      updateCgnStatusActivityInput
+    );
+
+    const updateCgnResult = ActivityResult.decode(
+      updateStatusResult
+    ).getOrElseL(e =>
+      trackExAndThrow(e, "cgn.update.exception.decode.activityOutput")
+    );
+
+    if (updateCgnResult.kind !== "SUCCESS") {
+      trackExAndThrow(
+        new Error("Cannot update CGN Status"),
+        "cgn.update.exception.failure.activityOutput"
+      );
+    }
+
+    const hasSendMessageActivity = [
+      RevokedStatusEnum.REVOKED.toString()
+    ].includes(newStatus.status);
```

Comment:
what is this code supposed to do ? can you add a comment (is not very clear from the context)

it looks like 

```suggestion
    const hasSendMessageActivity = 
      RevokedStatusEnum.REVOKED === newStatus.status;
```

would suffice

Reply:
It has been intended to contain all CGN statuses that must send a message with the infos about status change. At the moment it's only the `REVOKED` status. Later I will add also `CANCELED` and `ACTIVATED` statuses :)

Reply:
I will change it for now :)